types/webmidi/index.d.ts

```diff
@@ -4,193 +4,234 @@
-     * A Uint8Array containing the MIDI data bytes of a single MIDI message.
-     */
-    data: Uint8Array;
-  }
-
-  interface MIDIConnectionEvent extends Event {
-    /**
-     * The port that has been connected or disconnected.
-     */
-    port: MIDIPort;
-  }
+    requestMIDIAccess(options?: WebMidi.MIDIOptions): Promise<WebMidi.MIDIAccess>;
+}
 
-  interface MIDIConnectionEventInit extends EventInit {
-    /**
-     * The port that has been connected or disconnected.
-     */
-    port: MIDIPort;
-  }
+declare namespace WebMidi {
+    interface MIDIOptions {
+        /**
+         * This member informs the system whether the ability to send and receive system
+         * exclusive messages is requested or allowed on a given MIDIAccess object.
+         */
+        sysex: boolean;
+    }
+
+    /**
+     * This is a maplike interface whose value is a MIDIInput instance and key is its
+     * ID.
+     */
+    type MIDIInputMap = Map<string, MIDIInput>;
+
+    /**
+     * This is a maplike interface whose value is a MIDIOutput instance and key is its
+     * ID.
+     */
+    type MIDIOutputMap = Map<string, MIDIOutput>;
+
+    interface MIDIAccess extends EventTarget {
+        /**
+         * The MIDI input ports available to the system.
+         */
+        inputs: MIDIInputMap;
+
+        /**
+         * The MIDI output ports available to the system.
+         */
+        outputs: MIDIOutputMap;
+
+        /**
+         * The handler called when a new port is connected or an existing port changes the
+         * state attribute.
+         */
+        onstatechange(e: MIDIConnectionEvent): void;
+
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+
+        /**
+         * This attribute informs the user whether system exclusive support is enabled on
+         * this MIDIAccess.
+         */
+        sysexEnabled: boolean;
+    }
+
+    type MIDIPortType = 'input' | 'output';
+
+    type MIDIPortDeviceState = 'disconnected' | 'connected';
+
+    type MIDIPortConnectionState = 'open' | 'closed' | 'pending';
+
+    interface MIDIPort extends EventTarget {
+        /**
+         * A unique ID of the port. This can be used by developers to remember ports the
+         * user has chosen for their application.
+         */
+        id: string;
+
+        /**
+         * The manufacturer of the port.
+         */
+        manufacturer?: string;
+
+        /**
+         * The system name of the port.
+         */
+        name?: string;
+
+        /**
+         * A descriptor property to distinguish whether the port is an input or an output
+         * port.
+         */
+        type: MIDIPortType;
```

Comment:
Same question as above -- can't this just be removed? (Might be nice to move the docstring to both implementors though.)

Reply:
If this was removed, you wouldn't be able to access 'type' on e.g. MIDIConnectionEvent::port, which wouldn't make much sense.